File: sdk/python/inkbox/_cookies.py
"""
inkbox/_cookies.py

Shared cookie parsing and matching helpers for the sync SDK transports.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import timezone
from email.utils import parsedate_to_datetime
from typing import Mapping, Sequence
from urllib.parse import urlparse
import time


@dataclass
class Cookie:
    name: str
    value: str
    domain: str
    host_only: bool
    path: str
    secure: bool
    expires_at: float | None


class CookieJar:
    def __init__(self) -> None:
        self._cookies: dict[tuple[str, str, str], Cookie] = {}

    def header_for_url(self, url: str) -> str | None:
        parsed = urlparse(url)
        host = (parsed.hostname or "").lower()
        path = parsed.path or "/"
        is_secure = parsed.scheme == "https"
        now = time.time()

        pairs: list[str] = []
        expired: list[tuple[str, str, str]] = []
        for key, cookie in self._cookies.items():
            if cookie.expires_at is not None and cookie.expires_at <= now:
                expired.append(key)
                continue
            if cookie.secure and not is_secure:
                continue
            if cookie.host_only:
                if host != cookie.domain:
                    continue
            elif not _domain_matches(host, cookie.domain):
                continue
            if not _path_matches(path, cookie.path):
                continue
            pairs.append(f"{cookie.name}={cookie.value}")

        for key in expired:
            self._cookies.pop(key, None)

        return "; ".join(pairs) if pairs else None

    def store_from_headers(self, url: str, headers: Mapping[str, str] | object) -> None:
        for raw_cookie in _get_set_cookie_headers(headers):
            cookie = _parse_set_cookie(url, raw_cookie)
            if cookie is None:
                continue
            key = (cookie.domain, cookie.path, cookie.name)
            if cookie.expires_at is not None and cookie.expires_at <= time.time():
                self._cookies.pop(key, None)
                continue
            self._cookies[key] = cookie


def _get_set_cookie_headers(headers: Mapping[str, str] | object) -> Sequence[str]:
    get_list = getattr(headers, "get_list", None)
    if callable(get_list):
        return get_list("set-cookie")

    get = getattr(headers, "get", None)
    if callable(get):
        value = get("set-cookie")
        return [value] if value else []

    return []


def _parse_set_cookie(url: str, header: str) -> Cookie | None:
    parts = [part.strip() for part in header.split(";") if part.strip()]
    if not parts:
        return None

    name, sep, value = parts[0].partition("=")
    if not sep or not name:
        return None

    parsed = urlparse(url)
    domain = (parsed.hostname or "").lower()
    host_only = True
    path = _default_path(parsed.path or "/")
    secure = False
    expires_at: float | None = None

    for attr in parts[1:]:
        key, _, attr_value = attr.partition("=")
        key = key.strip().lower()
        attr_value = attr_value.strip()
        if key == "domain" and attr_value:
            domain = attr_value.lstrip(".").lower()
            host_only = False
        elif key == "path" and attr_value.startswith("/"):
            path = attr_value
        elif key == "secure":
            secure = True
        elif key == "max-age":
            try:
                expires_at = time.time() + int(attr_value)
            except ValueError:
                pass
        elif key == "expires" and attr_value:
            try:
                expires_at = parsedate_to_datetime(attr_value).astimezone(tz=timezone.utc).timestamp()
            except (TypeError, ValueError, IndexError):
                pass

    return Cookie(
        name=name,
        value=value,
        domain=domain,
        host_only=host_only,
        path=path,
        secure=secure,
        expires_at=expires_at,
    )


def _default_path(path: str) -> str:
    if not path or not path.startswith("/"):
        return "/"
    if path == "/":
        return "/"
    return path.rsplit("/", 1)[0] or "/"


def _domain_matches(host: str, domain: str) -> bool:
    return host == domain or host.endswith(f".{domain}")


def _path_matches(request_path: str, cookie_path: str) -> bool:
    if request_path == cookie_path:
        return True
    if request_path.startswith(cookie_path):
        return cookie_path.endswith("/") or request_path[len(cookie_path)] == "/"
    return False

File: sdk/rust/src/cookies.rs
//! Shared cookie parsing and matching for the blocking transport.
//!
//! A faithful port of `inkbox/_cookies.py`: a minimal RFC-6265-ish jar used so
//! the tunnels CSR-signing flow can carry server-set cookies across the
//! multi-request handshake. Not a general-purpose cookie store.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use url::Url;

#[derive(Debug, Clone)]
struct Cookie {
    name: String,
    value: String,
    domain: String,
    host_only: bool,
    path: String,
    secure: bool,
    /// Unix seconds; `None` means session cookie (no expiry).
    expires_at: Option<f64>,
}

/// Thread-safe cookie jar keyed by `(domain, path, name)`, matching the Python
/// dict keying so duplicate cookies overwrite rather than accumulate.
#[derive(Debug, Default)]
pub struct CookieJar {
    cookies: Mutex<HashMap<(String, String, String), Cookie>>,
}

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

impl CookieJar {
    pub fn new() -> Self {
        Self::default()
    }

    /// Build the `Cookie:` header value for a request URL, dropping expired
    /// entries as a side effect. Returns `None` when nothing matches.
    pub fn header_for_url(&self, url: &str) -> Option<String> {
        let parsed = Url::parse(url).ok()?;
        let host = parsed.host_str().unwrap_or("").to_lowercase();
        let path = if parsed.path().is_empty() {
            "/"
        } else {
            parsed.path()
        };
        let is_secure = parsed.scheme() == "https";
        let now = now_secs();

        let mut jar = self.cookies.lock().unwrap();
        let mut expired: Vec<(String, String, String)> = Vec::new();
        let mut pairs: Vec<String> = Vec::new();

        for (key, cookie) in jar.iter() {
            if let Some(exp) = cookie.expires_at {
                if exp <= now {
                    expired.push(key.clone());
                    continue;
                }
            }
            if cookie.secure && !is_secure {
                continue;
            }
            if cookie.host_only {
                if host != cookie.domain {
                    continue;
                }
            } else if !domain_matches(&host, &cookie.domain) {
                continue;
            }
            if !path_matches(path, &cookie.path) {
                continue;
            }
            pairs.push(format!("{}={}", cookie.name, cookie.value));
        }

        for key in expired {
            jar.remove(&key);
        }

        if pairs.is_empty() {
            None
        } else {
            Some(pairs.join("; "))
        }
    }

    /// Ingest every `Set-Cookie` header from a response.
    pub fn store_from_headers<'a>(
        &self,
        url: &str,
        set_cookie_values: impl Iterator<Item = &'a str>,
    ) {
        for raw in set_cookie_values {
            if let Some(cookie) = parse_set_cookie(url, raw) {
                let key = (
                    cookie.domain.clone(),
                    cookie.path.clone(),
                    cookie.name.clone(),
                );
                if let Some(exp) = cookie.expires_at {
                    if exp <= now_secs() {
                        self.cookies.lock().unwrap().remove(&key);
                        continue;
                    }
                }
                self.cookies.lock().unwrap().insert(key, cookie);
            }
        }
    }
}

fn parse_set_cookie(url: &str, header: &str) -> Option<Cookie> {
    let parts: Vec<&str> = header
        .split(';')
        .map(|p| p.trim())
        .filter(|p| !p.is_empty())
        .collect();
    let first = parts.first()?;
    let (name, value) = first.split_once('=')?;
    if name.is_empty() {
        return None;
    }

    let parsed = Url::parse(url).ok()?;
    let mut domain = parsed.host_str().unwrap_or("").to_lowercase();
    let mut host_only = true;
    let mut path = default_path(if parsed.path().is_empty() {
        "/"
    } else {
        parsed.path()
    });
    let mut secure = false;
    let mut expires_at: Option<f64> = None;

    for attr in &parts[1..] {
        let (k, v) = match attr.split_once('=') {
            Some((k, v)) => (k.trim().to_lowercase(), v.trim().to_string()),
            None => (attr.trim().to_lowercase(), String::new()),
        };
        match k.as_str() {
            "domain" if !v.is_empty() => {
                domain = v.trim_start_matches('.').to_lowercase();
                host_only = false;
            }
            "path" if v.starts_with('/') => path = v,
            "secure" => secure = true,
            "max-age" => {
                if let Ok(secs) = v.parse::<i64>() {
                    expires_at = Some(now_secs() + secs as f64);
                }
            }
            "expires" if !v.is_empty() => {
                if let Some(ts) = parse_http_date(&v) {
                    expires_at = Some(ts);
                }
            }
            _ => {}
        }
    }

    Some(Cookie {
        name: name.to_string(),
        value: value.to_string(),
        domain,
        host_only,
        path,
        secure,
        expires_at,
    })
}

/// Best-effort RFC 1123 / asctime date parse to a Unix timestamp. The Python
/// SDK leans on `email.utils.parsedate_to_datetime`; we only need enough to
/// honour `Expires` on the handful of cookies the control plane sets.
fn parse_http_date(s: &str) -> Option<f64> {
    // Format: "Wdy, DD Mon YYYY HH:MM:SS GMT"
    let s = s.trim();
    let comma = s.find(',')?;
    let rest = s[comma + 1..].trim();
    let mut it = rest.split_whitespace();
    let day: i64 = it.next()?.parse().ok()?;
    let mon = month_num(it.next()?)?;
    let year: i64 = it.next()?.parse().ok()?;
    let time = it.next()?;
    let mut tparts = time.split(':');
    let hh: i64 = tparts.next()?.parse().ok()?;
    let mm: i64 = tparts.next()?.parse().ok()?;
    let ss: i64 = tparts.next()?.parse().ok()?;
    Some(days_from_civil(year, mon, day) as f64 * 86400.0 + (hh * 3600 + mm * 60 + ss) as f64)
}

fn month_num(m: &str) -> Option<i64> {
    Some(match m {
        "Jan" => 1,
        "Feb" => 2,
        "Mar" => 3,
        "Apr" => 4,
        "May" => 5,
        "Jun" => 6,
        "Jul" => 7,
        "Aug" => 8,
        "Sep" => 9,
        "Oct" => 10,
        "Nov" => 11,
        "Dec" => 12,
        _ => return None,
    })
}

/// Days since the Unix epoch for a civil date (Howard Hinnant's algorithm).
fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = y - era * 400;
    let doy = (153 * (if m > 2 { m - 3 } else { m + 9 }) + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

fn default_path(path: &str) -> String {
    if !path.starts_with('/') {
        return "/".to_string();
    }
    if path == "/" {
        return "/".to_string();
    }
    match path.rsplit_once('/') {
        Some((head, _)) if !head.is_empty() => head.to_string(),
        _ => "/".to_string(),
    }
}

fn domain_matches(host: &str, domain: &str) -> bool {
    host == domain || host.ends_with(&format!(".{domain}"))
}

fn path_matches(request_path: &str, cookie_path: &str) -> bool {
    if request_path == cookie_path {
        return true;
    }
    if request_path.starts_with(cookie_path) {
        return cookie_path.ends_with('/')
            || request_path.as_bytes().get(cookie_path.len()) == Some(&b'/');
    }
    false
}
